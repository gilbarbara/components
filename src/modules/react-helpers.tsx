import {
  Children,
  cloneElement,
  ComponentProps,
  ElementType,
  FC,
  isValidElement,
  ReactElement,
  ReactNode,
} from 'react';
import { PlainObject } from '@gilbarbara/types';

export function enhanceChildren<T extends ElementType>(
  children: ReactNode,
  type: T,
  props: PlainObject,
): ReactNode {
  const getProps = (includeProps: boolean, nodes: ReactNode) => {
    const enhancedProps: PlainObject = includeProps ? { ...props } : {};

    if (nodes) {
      enhancedProps.children = nodes;
    }

    return enhancedProps;
  };

  return Children.map(children, child => {
    if (!isValidElement(child)) {
      return child; // Return non-element nodes unchanged
    }

    const enhancedChildren = child.props.children
      ? enhanceChildren(child.props.children, type, props) // Recursively enhance children
      : undefined;

    // Enhance the target component by merging props
    if (child.type === type) {
      return cloneElement(child, getProps(true, enhancedChildren));
    }

    // Return other elements unchanged, but with enhanced nested children
    return cloneElement(child, getProps(false, enhancedChildren));
  });
}

export function getMatchingChildren<T extends ElementType>(
  children: ReactNode,
  type: T,
): { elements: ReactElement<ComponentProps<T>>[]; hasInvalidElements: boolean } {
  const matchingElements: ReactElement<ComponentProps<T>>[] = [];
  let hasInvalidElements = false;

  const processChildren = (nodes: ReactNode) => {
    Children.forEach(nodes, child => {
      if (!isValidElement(child)) {
        hasInvalidElements = true;
        // eslint-disable-next-line no-console
        console.warn('Invalid child detected: Not a React element.', child);

        return;
      }

      // Check if the child matches the valid component type
      if (child.type === type) {
        matchingElements.push(child as ReactElement<ComponentProps<T>>);

        return;
      }

      // If the child is a functional component, render it and process its output
      if (typeof child.type === 'function') {
        const renderedOutput = (child.type as FC)(child.props);

        processChildren(renderedOutput);

        return;
      }

      // Recursively process nested children if available
      const { children: nestedChildren } = child.props || {};

      if (nestedChildren) {
        processChildren(nestedChildren);
      }
    });
  };

  processChildren(children);

  return { elements: matchingElements, hasInvalidElements };
}

export function splitReactChildren<T extends ElementType>(
  children: ReactNode,
  targetChild: T,
): [ReactNode, ReactElement<ComponentProps<T>>[]] {
  const matchingElements: ReactElement<ComponentProps<T>>[] = [];

  const withoutTargetChildren = Children.map(children, item => {
    if (!isValidElement(item)) {
      return item;
    }

    if (item.type === targetChild) {
      matchingElements.push(item as ReactElement<ComponentProps<T>>);

      return null;
    }

    return item;
  })?.filter(Boolean);

  return [withoutTargetChildren, matchingElements];
}
