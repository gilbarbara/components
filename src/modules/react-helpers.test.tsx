import { PropsWithChildren } from 'react';
import { render, screen } from '@testing-library/react';

import { enhanceChildren, getMatchingChildren, splitReactChildren } from './react-helpers';

function ComponentA({ children, name }: PropsWithChildren<{ name?: string }>) {
  return <div data-testid="component-a">{name ?? children}</div>;
}

function ComponentB({ name = 'Not Enhanced' }: { name?: string }) {
  return <div data-testid="component-b">{name}</div>;
}

describe('enhanceChildren', () => {
  it('should enhance the matching components', () => {
    const children = enhanceChildren(
      <>
        Hello
        <ComponentA />
        <>
          <ComponentA />
          <ComponentB />
        </>
      </>,
      ComponentA,
      {
        name: 'Enhanced',
      },
    );

    render(children);

    const enhancedComponents = screen.getAllByTestId('component-a');

    expect(enhancedComponents).toHaveLength(2);
    enhancedComponents.forEach(instance => {
      expect(instance).toHaveTextContent('Enhanced');
    });

    // Assert ComponentB remains untouched
    const untouchedComponents = screen.getAllByTestId('component-b');

    expect(untouchedComponents).toHaveLength(1);
    untouchedComponents.forEach(instance => {
      expect(instance).toHaveTextContent('Not Enhanced');
    });
  });
});

describe('getMatchingChildren', () => {
  beforeAll(() => {
    vi.spyOn(console, 'warn').mockImplementation(() => {});
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('should get the matching children', () => {
    const { elements, hasInvalidElements } = getMatchingChildren(
      <>
        <ComponentA>One</ComponentA>
        <ComponentA>Two</ComponentA>
        <ComponentB />
      </>,
      ComponentA,
    );

    const { container } = render(elements);

    expect(container.innerHTML).toMatchSnapshot();
    expect(hasInvalidElements).toBe(true);
  });
});

describe('splitReactChildren', () => {
  it('should return the splitted children', () => {
    const [otherChildren, selected] = splitReactChildren(
      ['Hello', <ComponentA />, <ComponentA>Two</ComponentA>, <ComponentB />],
      ComponentA,
    );

    expect(otherChildren).toHaveLength(2);
    expect(selected).toHaveLength(2);
    expect(selected[0]).toStrictEqual(<ComponentA />);
  });
});
