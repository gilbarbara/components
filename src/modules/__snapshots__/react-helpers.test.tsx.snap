// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getMatchingChildren > should get the matching children 1`] = `"<div data-testid="component-a">One</div><div data-testid="component-a">Two</div>"`;
